### ISSUE TITLE
Cannot get silos work

### ISSUE THREAD

**Issue by rbenit68** [protesilaos/denote#127]
Hi!. Thank you for denote!

The issue so far:

(setq denote-directory "~/orgfiles/denote")

~/orgfiles/2023/.dir-locals.el 
((nil . ((denote-directory . default-directory))))

Trying to complete a note in ~/orgfiles/2023:
M-x denote-subdirectory  RET
Select subdirectory: ~/orgfiles/2023 RET [No match]

Could you please explain in detail what I am doing wrong or how to use properly silos? Thanks.
Emacs 28.1 MacOS / denote commits on Jan 5, 2023

**Comment by aa-parky** [comment]
I maybe wrong, but your 2023 is not a subdirectory of your default denote directory, rather both those directories are subdirectories of your orgfiles directory.

**Comment by rbenit68** [comment]
Thanks, aa-parky.

I have followed a scheme similar to proposed in manual:

;; This is the global value of 'denote-directory' (no need for a .dir-locals.el)
~/Documents/notes
|-- 20210303T120534--this-is-a-test__journal_philosophy.txt
|-- 20220303T120534--another-sample__journal_testing.md
`-- 20220620T181255--the-third-test__keyword.org

;; A silo with notes for the employer
~/different/path/to/notes-for-employer
|-- .dir-locals.el
|-- 20210303T120534--this-is-a-test__conference.txt
|-- 20220303T120534--another-sample__meeting.md
`-- 20220620T181255--the-third-test__keyword.org

;; Another silo with notes for my volunteering
~/different/path/to/notes-for-volunteering
|-- .dir-locals.el
|-- 20210303T120534--this-is-a-test__activism.txt
|-- 20220303T120534--another-sample__teambuilding.md
`-- 20220620T181255--the-third-test__keyword.org

So should I assume that cannot be ~/different/path/to/notes-for-employer and ~/different/path/to/notes-for-volunteering actually, but subdirectories of ~/Documents/notes instead?

**Comment by MirkoHernandez** [comment]
Your directory setup seems right to me. 

`denote-subdirectory`  requires that a directory should  already be  created.  Create the directories that you need in your silo and then you can fill the directory with notes using `denote-subdirectory`.

**Comment by protesilaos** [comment]
@rbenit68 Indeed the `denote-subdirectory` does not create subdirectories.  Please try what @MirkoHernandez suggests.  Originally I also got the same impression as @aa-parky.

* * *

On this note, does anyone think we should revise how we handle this?  Maybe it makes sense to create a subdirectory if the user's input does not match one of the candidates?  I avoided this before for two reasons:

1. Typos happen and polluting the filesystem in this way may be annoying.
2. Some completion UIs will default to the topmost completion candidate, so we may get false positives.

Basically, I designed that way to be simple.  But I am happy to read about alternatives.

**Comment by rbenit68** [comment]
Thank you @protesilaos.
I am aware that `denote-subdirectory` does not create subdirectories, the fact is that directories are already created by the function do not offer them as choices.

I think the obstacle is that `denote-subdirectory-prompt` calls `denote-directory-subdirectories`, and this function "Return list of subdirectories in variable `denote-directory`." So I have to assume than silos have to be subdirectories of `denote-directory`

Shouldn't manual read:

`There is, however, the possibility to maintain separate subdirectories [instead of directories] of notes. By “separate”, we mean that they do not communicate with each other: no linking between them, no common keywords, nothing. Think of the scenario where one set of notes is for private use and another is for an employer. We call these separate subdirectories [instead of directories] “silos”.`

`To create silos, the user must specify a local variable at the root of the desired subdirectory [instead of directory]. This is done by creating a .dir-locals.el file, with the following contents: `

Best.

**Comment by protesilaos** [comment]
Hello @rbenit68!

Silos do not need to be subdirectories of the default `denote-directory`.  For example, I have my `~/Documents/notes/` as my main directory and `~/Videos/recordings/` as a "silo".

When we set the `.dir-locals.el` we change the value of the `denote-directory` locally.  So any command that reads that variable will adapt its behaviour accordingly.  In practice, this means that it depends from _where_ you call the `denote-subdirectory` command (or any other command).

To use my example, if I am in `~/Videos/recordings/` (e.g. with Dired) and do `M-x denote` it will create the note in the videos' directory.  But if I am anywhere else, the same command will read the global value of `denote-directory` and thus put the new note in `~/Documents/notes/`.

Does this make sense?

**Comment by rbenit68** [comment]
Thank you @protesilaos.

Solved! Now I see the intended behaviour.

I encourage you to include the paragraph in the manual:

`(...) **if I am in ~/Videos/recordings/ (e.g. with Dired)** and do M-x denote it will create the note in the videos' directory. But if I am anywhere else, the same command will read the global value of denote-directory and thus put the new note in ~/Documents/notes/.`

Thanks again. For me the issue can be closed. Best.

**Comment by protesilaos** [comment]
@rbenit68 Thank you for the suggestion!  I can add it, though please let me know what you think of the following quote?  It is already in the manual, but maybe I need to make it more clear?

> When inside the directory that contains this `.dir-locals.el` file, all Denote commands/functions for note creation, linking, the inference of available keywords, et cetera will use the silo as their point of reference.  They will not read the global value of `denote-directory`.  The global value of `denote-directory` is read everywhere else except the silos.

**Comment by rbenit68** [comment]
Thank you @protesilaos.

I am aware of the paragraph you mention and yes, it does explain the procedure for putting notes in the silo. The fact is that when I started to get interested in silos, I couldn't clearly find a function that in addition to asking for the title of the note, asked for the directory where to put it. In a closed issue I found a reference to the `denote-subdirectory` function and I started to insist on that way.

In the paragraph you quote it is clearly stated (my bad!) that

>  When inside the directory that contains this `.dir-locals.el` file, all Denote commands/functions for note creation, linking, the inference of available keywords, et cetera will use the silo as their point of reference. 

However, I could not imagine that to put a note in a silo I would have to select the directory outside the denote functions. I am not used to use dired and as I said before I was looking for a function that would ask for the directory of the silo.

Finally, I would like to say that what I have missed in the manual is a sequence of functions on how to put a note in a silo. I think it should be made clear that the functions are the same, quoting them explicitly, as well as some usual procedures, interactive or by defining other functions, on how to change directory before using the denote functions, e.g. using dired or with elisp in the form of 

```
(let* ((denote-directory "~/orgfiles/denote/2023"))
  (call-interactively 'denote))
```

Best.

**Comment by MirkoHernandez** [comment]
@rbenit68  I' think that I understand now what you were trying  to achieve with the `denote-subdirectory ` function.  The denote commands try to find the current `denote-directory` depending on the context in which they are called. In dired or in an active buffer the commands use the  `.dir-locals.el`  variable if it is set. In other contexts (a source code file for example)  the original `denote-directory` will be used.

Maybe something like this would help for your use case.

```
(let* (silos '("dir1" "dir2"  "dir3")
        (denote-directory (completing-read "Silo:" silos))
(call-interactively 'denote))  
```

**Comment by rbenit68** [comment]
Thank you @MirkoHernandez.

I appreciate the comments and the improved code you share. It would be very useful indeed.

Best.

**Comment by protesilaos** [comment]
@rbenit68 Here is the idea that @MirkoHernandez proposed:

```elisp
(defvar my-denote-silo-directories
  `("/home/prot/Videos/recordings"
    "/home/prot/Documents/books"
    ;; You don't actually need to include the `denote-directory' here
    ;; if you use the regular commands in their global context.  I am
    ;; including it for completeness.
    ,denote-directory)
  "List of file paths pointing to my Denote silos.")

(defvar my-denote-commands-for-silos
  '(denote
    denote-date
    denote-subdirectory
    denote-template
    denote-type)
  "List of Denote commands to call after selecting a silo.")

(defun my-denote-pick-silo-then-command (silo command)
  "Select SILO and run Denote COMMAND in it."
  (interactive
   (list (completing-read "Select a silo: " my-denote-silo-directories nil t)
         (intern (completing-read "Run command in silo: " my-denote-commands-for-silos nil t))))
  (let ((denote-directory silo))
    (call-interactively command)))
```

Edit the `my-denote-silo-directories` to specify your desired filesystem paths and then try `M-x my-denote-pick-silo-then-command`.  Depending on your workflow, you can skip the selection of the command and always run the same one.  I can show you how, if needed.

**Comment by rbenit68** [comment]
Thank you @protesilaos. It is much more than I needed. I really appreciate your (Mirko's and yours) suggestions and offer.

Best.

**Comment by protesilaos** [comment]
Hello again folks!  I just expanded the manual to cover this use-case:

```
commit e9c2e8d73c98791536b590378ecefa9f0546eee5
Author: Protesilaos Stavrou <info@protesilaos.com>
Date:   Mon Jan 16 07:01:35 2023 +0200

    Add documentation to select silo interactively
    
    I am providing this in response to a request from GitHub user
    rbenit68.  The discussion took place in issue 127 on the GitHub
    mirror, with the participation of Mirko Hernandez:
    <https://github.com/protesilaos/denote/issues/127>.
    
    The custom code I provide is the expanded version of an idea put forth
    by Mirko, to whom I am thankful.

 README.org | 63 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-
 1 file changed, 62 insertions(+), 1 deletion(-)

```

We can safely close this now.  Thank you!
